### PR TITLE
SmtpMailer: add support for AUTH PLAIN

### PR DESCRIPTION
**Pull request by grongor**
It's strange that there are some SMTP servers not supporting the LOGIN mechanism but I stumbled upon one ... so here is the fix :)

### CODE REVIEW

**Comment by JanTvrdik** [comment]
I wonder how did you manage accidentally set executable flag on the file :smile:

**Comment by JanTvrdik** [comment]
https://tools.ietf.org/html/rfc2595#section-2.3 may be related.

File: src/Mail/SmtpMailer.php [hunk 1]
```diff
@@ -145,9 +146,22 @@ protected function connect()
 		}
 
 		if ($this->username != NULL && $this->password != NULL) {
-			$this->write('AUTH LOGIN', 334);
-			$this->write(base64_encode($this->username), 334, 'username');
-			$this->write(base64_encode($this->password), 235, 'password');
+			$authMechanisms = [];
+			foreach (preg_split("~\r?\n~", $ehloResponse) as $line) {
+				if (strpos($line, '250-AUTH ') === 0) {
```
Comment (JanTvrdik) [inline src/Mail/SmtpMailer.php]: shouldn't this be `250 AUTH`?

**Comment by grongor** [comment]
... some Windows magic I guess, I don't usually push code on this computer so it got a bit clumsy :D I will try to fix it ...

How should the STARTTLS be related? If it is required and for some reason it fails to activate then the execution stops - the credentials are never sent.

The space after return code is used to indicate the last line of the response. The link you posted is a good example of that.

**Comment by grongor** [comment]
Thanks for pointing out the space vs dash issue - I updated the implementation to count with both of them (because the AUTH can also be the last line of the response).

File: src/Mail/SmtpMailer.php [hunk 2]
```diff
@@ -145,9 +146,19 @@ protected function connect()
...
 		if ($this->username != NULL && $this->password != NULL) {
-			$this->write('AUTH LOGIN', 334);
-			$this->write(base64_encode($this->username), 334, 'username');
-			$this->write(base64_encode($this->password), 235, 'password');
+			$authMechanisms = [];
+			if (preg_match('~^250[ -]AUTH (.*)$~im', $ehloResponse, $matches)) {
+				$authMechanisms = explode(' ', trim($matches[1]));
+			}
+
+			if (array_search('PLAIN', $authMechanisms, true) !== false) {
```
Comment (JanTvrdik) [inline src/Mail/SmtpMailer.php]: can be simplified with `in_array`

---

Also now it uses PLAIN auth whenever possible, shouldn't we prefer LOGIN auth when both plain and login auth are allowed?

Comment (grongor) [inline src/Mail/SmtpMailer.php]: Alright, I will change it to in_array.

I also though about it and I don't think that there is a reason to prefer LOGIN mechanism over PLAIN. They both offer same security. The PLAIN mechanism will be just a bit faster ... correct me if I'm wrong.

**Comment by dg** [comment]
Thanks

**Pull request merged**